Add analytic data for a plane wave in Minkowski background

prayush opened the pull request:
## Proposed changes

Add analytic data class for a plane wave propagating in Minkowski spacetime. This specialization will provide initial data for the curved scalar-wave system to produce evolutions that can be compared directly with the flat scalar-wave system.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

prayush committed:
Add analytic data for plane wave in Minkowski background

wthrowe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ add_spectre_library(${LIBRARY})
   ${LIBRARY}
   PRIVATE
+  PlaneWaveMinkowski.cpp
   PlaneWave3DKerrSchild.cpp
```
'3' < 'M' and again below.

prayush replied:
fixed

wthrowe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PlaneWaveMinkowski.hpp:
```diff
@@ -0,0 +1,9 @@
+
+#pragma once
+
+#include "PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.tpp"
```
Move to cpp.

prayush replied:
moved. But all other analytic data instantiations in this directory adhere to this include format, i.e. include all `.tpp` and `.hpp` in the Data.hpp file, and only include `Data.hpp` in the `Data.cpp`. Should I change those too? See links [below](https://github.com/sxs-collaboration/spectre/pull/3100#discussion_r657835296)

wthrowe replied:
Yes.  Unnecessarily including a tpp, particularly in a header, is very bad.  tpp files are basically hpp files defining templates, but indicating that including them may be very compile-time expensive.

You do still need `ScalarWaveGr.hpp` for your intent for this file mentioned below.

wthrowe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PlaneWaveMinkowski.hpp:
```diff
@@ -0,0 +1,9 @@
+// Distributed under the MIT License.
```
You don't use this header anywhere.  Is there a reason to have it rather than having people include the Minkowsi.hpp and PlaneWave.hpp headers themselves?  (Having the cpp for the explicit instantiations makes sense.  I'm only questioning the hpp here.)

prayush replied:
The answer is consolidation of includes, I guess. That and it saves people remembering which two includes need to be made when using this data. The same choice has been made for all analytic data instantiations in here here, e.g. [`PlaneWave3DKerrSchild`](https://github.com/sxs-collaboration/spectre/blob/develop/src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PlaneWave3DKerrSchild.cpp), [`RegularSphericalWave3DKerrSchild`](https://github.com/sxs-collaboration/spectre/blob/develop/src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/RegularSphericalWave3DKerrSchild.cpp).

wthrowe replied:
But you're not using it, you're including all the pieces separately.  If you want this file to be the thing that is included when you want to use the class, you should include it when you use the class.

Honestly, if I were using `ScalarWaveGr<ScalarWave::Solutions::PlaneWave<3>, gr::Solutions::Minkowski<3>>`, I would find it easier to remember to include `ScalarWaveGr.hpp`, `PlaneWave.hpp`, and `Minkowski.hpp` than `PlaneWaveMinkowski.hpp`.  (As is done in the test.)

This file might be useful if it also defined an alias for this class so people didn't have to know exactly which analytic solutions are used internally.  If we had something like `using PlaneWaveMinkowski = ScalarWaveGr<...>;` so that it hid the classes being used, then I think hiding the includes would make sense.

prayush replied:
I think I agree with you (and so does the unit test) that if I wanted to use the class `ScalarWaveGr<A, B>` I would intuitively include `ScalarWaveGr`, `A`, and `B` instead of some arbitrarily named header `ScalarWaveAB`. Later on I might find that such a header exists and replace these includes with it, but its definitely not intuitive. I also don't like having an alias for the same reason, a user may just find the flexibility of `ScalarWaveGr` much more easier to remember.

In that spirit, I have removed `PlaneWaveMinkowski.hpp` here. Going one step ahead, I have removed 2 other similar headars in this directory to uphold the new convention. Ironically enough, no other code was impacted by these headers' removal, clearly showing how useful they were!

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp:
```diff
@@ -210,6 +210,105 @@ void test_construct_from_options() noexcept {
+
+    // construct internals of data constructed from options
+    const gr::Solutions::Minkowski<3> flat_space_solution{};
+    const ScalarWave::Solutions::PlaneWave<3> flat_wave_solution{
```
This isn't used.  Can you move it below where you comment that you can't copy it?  Same for the other tests.

prayush replied:
removed them in fact, there's no use constructing these objects.

prayush committed:
Remove unnecessary headers in AnalyticData/CurvedWaveEqn

These headers were intended to hide the includes for
ScalarWaveGr<SW, Bkground> for different choices of SW and
Bkground. However, even the unit tests of these classes do
not use this compound header and instead include
ScalarWaveGr, SW, Bkground individually. This is the more
natural choice anyway, so this commit removes the
unnecessary compound headers.

prayush commented:
Hi @wthrowe - sorry for getting to this so late. I hope your concerns are satisfactorily addressed now. Let me know if I can squash these fixups.

wthrowe commented:
OK, all sounds good then.  Go ahead and squash.

prayush commented:
> OK, all sounds good then. Go ahead and squash.

Squashed and rebased on develop. Maybe this can be merged now @wthrowe ?

wthrowe reviewed (dismissed):
Needs a second review.  I'll solicit one.

markscheel commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp:
```diff
@@ -383,6 +473,26 @@ SPECTRE_TEST_CASE("Unit.AnalyticData.CurvedWaveEquation.ScalarWaveGr",
     test_kerr(curved_wave_data, bh_solution, flat_wave_solution);
     test_kerr_schild_vars(curved_wave_data, bh_solution);
   }
+  {  // test with wave profile = plane in flat space in 3D
```
[optional] It seems a bit strange to me that all the Kerr variables (mass, spin, bh_solution, and especially `background_tag` which is probably the easiest to use accidentally because it doesn't sound like a Kerr thing) are still in scope for this new flatspace test.  And none of those variables are used for this flatspace test.

Perhaps (as in many other tests) this test can be refactored into two functions: the first function is the test before this PR (which includes all of the blocks above), and the second function is this new test.  Then `SPECTRE_TEST_CASE("Unit.AnalyticData.CurvedWaveEquation.ScalarWaveGr",...` can be very short and simple, i.e. it just calls two functions.

prayush replied:
Thats fair, I've pushed a fixup splitting this test.

prayush replied:
@markscheel - can I squash this fix?

nilsvu replied:
I'm not a reviewer on this PR, so I let @wthrowe decide here, but we've been keeping the number of test cases low in favor of a single test case that calls into a bunch of functions. This is to reduce overhead from test-startup times. My suggestion is to change both test cases you have here into functions (comes down to a two-line change), and call them from a single test case. If you don't want to bother with that, I think you can also just drop the fixup commit. If you agree to make the change, just squash it into the existing fixup. Then we can get this merged once people in the US wake up :)

prayush replied:
Sure, I moved the two tests into functions and now call them from a single test case. This now satisfies both Mark's and your comments.

I've squashed this change with older fixups, and the PR is good to go now.

markscheel reviewed (commented):
Looks fine to me.  I had one optional comment that would make the test easier to understand.

nilsvu commented:
@markscheel please look at the fixup

markscheel reviewed (dismissed):
Please squash, looks good!

prayush commented:
Thanks @wthrowe , @markscheel for looking at this. Squashed all fixups, including to @nilsleiffischer 's last suggestion. This PR is ready to go.

wthrowe reviewed (approved):
I'm going to take @markscheel's last comment as an approval and merge this.
